File: scripts/live-smoke.mjs
#!/usr/bin/env node

// Drives the PUBLISHED SDK against the LIVE api.cognitum.one as an end user
// would, from a clean install off the public registry.
//
// Everything else in CI tests the SDK against mocks. That proves the client
// is internally consistent; it cannot notice the gateway changing a route, a
// response shape, or an auth rule underneath us. This is the only check that
// exercises the seam where the published artifact meets the deployed service.
//
// Assertions are on SEMANTICS, not status codes. A 200 carrying an empty
// completion is a failure worth catching; asserting `status === 200` would
// call it a pass.
//
// Usage: node live-smoke.mjs            (reads COGNITUM_API_KEY, COGNITUM_API_BASE_URL)

import { createRequire } from "node:module";

// The SDK is resolved from the CURRENT WORKING DIRECTORY, not from this
// file's location. This script lives in the repo while the package under test
// is installed into a throwaway directory elsewhere, so a static top-level
// `import "@cognitum-one/sdk/..."` resolves against scripts/ and dies with
// ERR_MODULE_NOT_FOUND no matter how correctly the package was published --
// the smoke test would report a failure that says nothing about the release.
// (Module resolution itself, ESM and CJS, is covered by
// smoke-published-package.mjs; this script's job is the network contract.)
const require = createRequire(`${process.cwd()}/`);
const { StaticApiKeyCredentialProvider } = require("@cognitum-one/sdk/agentic");
const { MetaLlmClient } = require("@cognitum-one/sdk/meta-llm");

const baseUrl = process.env.COGNITUM_API_BASE_URL ?? "https://api.cognitum.one";
const apiKey = process.env.COGNITUM_API_KEY;

function assert(condition, message) {
  if (!condition) throw new Error(message);
}

async function check(name, fn) {
  const started = Date.now();
  try {
    await fn();
    console.log(`ok   ${name} (${Date.now() - started}ms)`);
    return null;
  } catch (error) {
    console.error(`FAIL ${name}: ${error.message}`);
    return `${name}: ${error.message}`;
  }
}

function currentMonth() {
  // Not `new Date()` for cleanliness -- the usage window just needs to be a
  // month the account plausibly has data in, and "now" is the right one.
  const now = new Date();
  return `${now.getUTCFullYear()}-${String(now.getUTCMonth() + 1).padStart(2, "0")}`;
}

async function main() {
  assert(apiKey, "COGNITUM_API_KEY is not set");
  const client = new MetaLlmClient({
    baseUrl,
    credentialProvider: new StaticApiKeyCredentialProvider({
      apiKey,
      product: "meta-llm",
      normalizedOrigin: baseUrl,
      audience: baseUrl,
    }),
  });

  console.log(`live smoke against ${baseUrl}`);
  const failures = [];

  failures.push(await check("health reports healthy", async () => {
    const { data } = await client.health();
    assert(data?.status === "healthy", `expected status "healthy", got ${JSON.stringify(data?.status)}`);
  }));

  failures.push(await check("models lists the tier aliases", async () => {
    const { data } = await client.models();
    const ids = (data?.data ?? []).map((model) => model.id);
    assert(ids.length > 0, "model list is empty");
    for (const required of ["cognitum-low"]) {
      assert(ids.includes(required), `model list is missing ${required}; got ${ids.join(", ")}`);
    }
  }));

  failures.push(await check("whoami returns authenticated identity", async () => {
    const { data } = await client.whoami();
    assert(data && typeof data === "object", `identity response is empty: ${JSON.stringify(data)}`);
  }));

  failures.push(await check("chat.completions returns real content", async () => {
    const { data } = await client.chat.completions({
      model: "cognitum-low",
      messages: [{ role: "user", content: "Reply with exactly: ok" }],
      max_tokens: 8,
    });
    const content = data?.choices?.[0]?.message?.content;
    assert(typeof content === "string" && content.trim().length > 0, `empty completion content: ${JSON.stringify(data?.choices?.[0])}`);
    assert(data?.usage?.total_tokens > 0, "completion reported zero total_tokens -- usage accounting is not wired");
  }));

  failures.push(await check("messages.create returns real content", async () => {
    const { data } = await client.messages.create({
      model: "cognitum-low",
      messages: [{ role: "user", content: "Reply with exactly: ok" }],
      max_tokens: 8,
    });
    const text = data?.content?.find((block) => block.type === "text")?.text;
    assert(typeof text === "string" && text.trim().length > 0, `empty message content: ${JSON.stringify(data?.content)}`);
  }));

  failures.push(await check("usage returns non-empty monthly accounting", async () => {
    const month = currentMonth();
    const { data } = await client.usage({ from: month, to: month });
    assert(data?.totals?.totalTokens > 0, `usage reports zero or missing totalTokens: ${JSON.stringify(data?.totals)}`);
    assert(data.totals.requests > 0, "usage reports zero monthly requests");
  }));

  const real = failures.filter(Boolean);
  if (real.length > 0) {
    console.error(`\n${real.length} live check(s) failed:\n- ${real.join("\n- ")}`);
    process.exit(1);
  }
  console.log(`\nall live checks passed against ${baseUrl}`);
}

main().catch((error) => {
  console.error(error.stack ?? error.message);
  process.exit(1);
});

File: scripts/live_smoke.py
#!/usr/bin/env python3
"""Drive the published PyPI SDK against the live Cognitum gateway.

Run this only from a clean environment containing ``cognitum-sdk`` from
PyPI. Assertions deliberately check response meaning, not HTTP status alone.
"""

from __future__ import annotations

import asyncio
import os
from collections.abc import Awaitable, Callable
from datetime import datetime, timezone
from time import monotonic

from cognitum.agentic import StaticApiKeyCredentialProvider
from cognitum.meta_llm import (
    AnthropicMessageParam,
    AnthropicMessageRequest,
    ChatCompletionRequest,
    ChatMessage,
    MetaLlmClient,
    MetaLlmClientConfig,
    UsageQuery,
)


def require(condition: bool, message: str) -> None:
    if not condition:
        raise RuntimeError(message)


async def check(name: str, operation: Callable[[], Awaitable[None]]) -> str | None:
    started = monotonic()
    try:
        await operation()
        print(f"ok   {name} ({(monotonic() - started) * 1000:.0f}ms)")
        return None
    except Exception as error:  # noqa: BLE001 - aggregate all independent live checks
        print(f"FAIL {name}: {error}")
        return f"{name}: {error}"


async def main() -> None:
    base_url = os.getenv("COGNITUM_API_BASE_URL", "https://api.cognitum.one")
    api_key = os.getenv("COGNITUM_API_KEY")
    require(bool(api_key), "COGNITUM_API_KEY is not set")

    client = MetaLlmClient(
        MetaLlmClientConfig(
            base_url=base_url,
            credential_provider=StaticApiKeyCredentialProvider(
                product="meta-llm",
                normalized_origin=base_url,
                audience=base_url,
                api_key=api_key,
            ),
        )
    )
    print(f"live Python smoke against {base_url}")

    async def health() -> None:
        result = await client.health()
        require(result.data.status == "healthy", f"expected healthy, got {result.data.status!r}")

    async def models() -> None:
        result = await client.models()
        ids = [model.id for model in result.data.models]
        require(bool(ids), "model list is empty")
        require("cognitum-low" in ids, f"model list is missing cognitum-low; got {ids}")

    async def whoami() -> None:
        result = await client.whoami()
        identity = result.data
        require(
            bool(identity.account_id or identity.credential_type or identity.tenant_id),
            f"identity response has no identity fields: {identity!r}",
        )

    async def chat_completion() -> None:
        result = await client.chat.completions(
            ChatCompletionRequest(
                model="cognitum-low",
                messages=[ChatMessage(role="user", content="Reply with exactly: ok")],
                max_tokens=8,
            )
        )
        require(bool(result.data.choices), "completion has no choices")
        content = result.data.choices[0].message.content
        require(isinstance(content, str) and bool(content.strip()), f"empty completion: {content!r}")
        require(
            result.data.usage is not None and result.data.usage.total_tokens > 0,
            "completion reported zero or missing total_tokens",
        )

    async def messages_create() -> None:
        result = await client.messages.create(
            AnthropicMessageRequest(
                model="cognitum-low",
                messages=[AnthropicMessageParam(role="user", content="Reply with exactly: ok")],
                max_tokens=8,
            )
        )
        text = next(
            (block.get("text") for block in result.data.content if block.get("type") == "text"),
            None,
        )
        require(isinstance(text, str) and bool(text.strip()), f"empty message content: {result.data.content!r}")

    async def usage() -> None:
        month = datetime.now(timezone.utc).strftime("%Y-%m")
        result = await client.usage(UsageQuery(from_=month, to=month))
        require(
            result.data.totals.total_tokens is not None and result.data.totals.total_tokens > 0,
            "usage reports zero or missing total_tokens",
        )
        require(
            result.data.totals.requests is not None and result.data.totals.requests > 0,
            "usage reports zero monthly requests",
        )

    failures: list[str] = []
    for name, operation in (
        ("health reports healthy", health),
        ("models lists the tier aliases", models),
        ("whoami returns authenticated identity", whoami),
        ("chat.completions returns real content", chat_completion),
        ("messages.create returns real content", messages_create),
        ("usage returns non-empty monthly accounting", usage),
    ):
        failure = await check(name, operation)
        if failure:
            failures.append(failure)

    await client.aclose()
    if failures:
        raise RuntimeError(f"{len(failures)} live check(s) failed:\n- " + "\n- ".join(failures))
    print(f"\nall Python live checks passed against {base_url}")


if __name__ == "__main__":
    asyncio.run(main())
